CRM/Contact/BAO/GroupContactCache.php

```diff
@@ -618,19 +604,52 @@ public static function load(&$group, $force = FALSE) {
       $childrenIDs = explode(',', $group->children);
       foreach ($childrenIDs as $childID) {
         $contactIDs = CRM_Contact_BAO_Group::getMember($childID, FALSE);
-        //Unset each contact that is removed from the parent group
+        // Unset each contact that is removed from the parent group
         foreach ($removed_contacts as $removed_contact) {
           unset($contactIDs[$removed_contact]);
         }
+        if (empty($contactIDs)) {
+          // This child group has no contact IDs so we don't need to add them to
+          continue;
+        }
         $values = [];
         foreach ($contactIDs as $contactID => $dontCare) {
           $values[] = "({$groupID},{$contactID})";
         }
-
-        self::store([$groupID], $values);
+        $str = implode(',', $values);
+        CRM_Core_DAO::executeQuery("INSERT IGNORE INTO $tempTable (group_id, contact_id) VALUES $str");
       }
     }
 
+    // grab a lock so other processes don't compete and do the same query
+    $lock = Civi::lockManager()->acquire("data.core.group.{$groupID}");
+    if (!$lock->isAcquired()) {
+      // this can cause inconsistent results since we don't know if the other process
+      // will fill up the cache before our calling routine needs it.
+      // however this routine does not return the status either, so basically
+      // its a "lets return and hope for the best"
+      return;
+    }
+
+    // Don't call clearGroupContactCache as we don't want to clear the cache dates
+    // The will get updated by updateCacheTime() below and not clearing the dates reduces
+    // the chance that loadAll() will try and rebuild at the same time.
+    $clearCacheQuery = "
```

Comment:
Why is this back in here?

Reply:
I'm unsure why but maybe @mattwire can comment, however we only want the contacts that match the criteria in the cache so this seems to do it

Reply:
hmm we should only ever be calling 'load' when the cache has already been emptied - ie on demand check if cache is valid. If not purge (if required) and then call load

Reply:
The idea here is that actively clearing the cache is expensive and we don't need to do it in advance. Setting the cache to expired is enough because that will trigger a rebuild next time it is required.  But obviously we do need to clear it before re-populating so the DELETE query + INSERT query are next to each other to be as atomic as possible. As soon as the cache table is actually up to date we set the cache time so that any process accessing the cache knows that it is up to date and can be used.